distributed/worker.py

```diff
@@ -2233,6 +2238,47 @@ def _get_client(self, timeout=3):
                 assert self._client.status == 'running'
         return self._client
 
+    @gen.coroutine
+    def register_worker_environments(self, stream=None, environments=None):
+        result = {}
+        for name, env in environments.items():
+            if name not in self.environments:
+                try:
+                    env = pickle.loads(env)
+                    meets_condition = yield self.executor_submit('condition', env.condition)
+                    # TODO: Should this be in some kind of "while running" loop with
+                    # a timeout?
+                    if gen.is_future(meets_condition):
+                        meets_condition = meets_condition.result()
+                    if meets_condition:
+                        # TODO: This currently blocks. I expect that that is
+                        # not desirable, as a long setup task may look like a
+                        # worker that's become unresponsive.
+                        yield self.executor_submit('setup', env.setup)
```

Comment:
What if setup/condition are coroutines?  This might not function well.

Actually, I'm tempted to just run them locally as is without a separate thread.  If people care about non-blocking behavior then they should write coroutines.  I think that we can expect a bit of expertise from anyone using Environments.  Most of the time this probably won't be an issue.